# Problem Set 2: Linear Algebra (Solutions)

# Vector Operations

## Problem 1: Economic Quantities as Vectors

Suppose we have data on five countries. Their GDP values (in trillions
of USD) are:

$$
\text{gdp} = \begin{pmatrix} 25.5 \\ 18.0 \\ 4.2 \\ 4.1 \\ 3.1 \end{pmatrix}
$$

and their consumption shares of GDP are:

$$
\text{cshare} = \begin{pmatrix} 0.68 \\ 0.54 \\ 0.55 \\ 0.52 \\ 0.65 \end{pmatrix}
$$

**(a)** Create these two vectors in Julia. Compute the consumption
vector $c = \text{gdp} \odot \text{cshare}$ (element-wise
multiplication). What is the total consumption across all five
countries?

In [1]:
gdp = [25.5, 18.0, 4.2, 4.1, 3.1]
cshare = [0.68, 0.54, 0.55, 0.52, 0.65]

c = gdp .* cshare
println("Consumption vector: $c")
println("Total consumption: $(sum(c)) trillion USD")

Consumption vector: [17.34, 9.72, 2.3100000000000005, 2.1319999999999997, 2.015]
Total consumption: 33.517 trillion USD

**(b)** Compute the dot product $\text{gdp} \cdot \text{cshare}$. Verify
that it equals the sum of the consumption vector from part (a). Explain
economically why $\text{gdp} \cdot \text{cshare} = \sum_i c_i$.

In [1]:
dp = dot(gdp, cshare)
println("Dot product: $dp")
println("Sum of consumption vector: $(sum(c))")
println("Equal? $(dp ≈ sum(c))")

Dot product: 33.517
Sum of consumption vector: 33.517
Equal? true

The dot product
$\text{gdp} \cdot \text{cshare} = \sum_i \text{gdp}_i \times \text{cshare}_i = \sum_i c_i$.
Economically, each country’s consumption is its GDP times its
consumption share. The dot product computes exactly this weighted sum:
it multiplies each country’s GDP by its consumption share and adds up
the results, giving total consumption across all countries.

**(c)** Compute the Euclidean norm $\|\text{gdp}\|$ and the norm
$\|c\|$. Which is larger, and why must this always be the case when all
consumption shares are less than 1?

In [1]:
println("‖gdp‖ = $(norm(gdp))")
println("‖c‖   = $(norm(c))")

‖gdp‖ = 31.910969900647018
‖c‖   = 20.22611551929831

The GDP norm is larger. Since every consumption share satisfies
$0 < \text{cshare}_i < 1$, we have
$c_i = \text{cshare}_i \times \text{gdp}_i < \text{gdp}_i$ for each
country. Therefore $c_i^2 < \text{gdp}_i^2$ for all $i$, so
$\|c\|^2 = \sum c_i^2 < \sum \text{gdp}_i^2 = \|\text{gdp}\|^2$, and
hence $\|c\| < \|\text{gdp}\|$.

## Problem 2: Projections

The **projection** of vector $a$ onto vector $b$ is defined as:

$$
\text{proj}_b(a) = \frac{a \cdot b}{\|b\|^2} \, b
$$

This gives the component of $a$ that lies in the direction of $b$.

**(a)** Let $a = [3.0, 4.0]$ and $b = [1.0, 0.0]$. Compute
$\text{proj}_b(a)$ by hand and verify using Julia.

In [1]:
a = [3.0, 4.0]
b = [1.0, 0.0]

proj = (dot(a, b) / norm(b)^2) * b
println("proj_b(a) = $proj")

proj_b(a) = [3.0, 0.0]

By hand: $a \cdot b = 3 \times 1 + 4 \times 0 = 3$ and $\|b\|^2 = 1$, so
$\text{proj}_b(a) = 3 \times [1, 0] = [3, 0]$. The projection onto the
$x$-axis simply picks out the $x$-component of $a$.

**(b)** Now let $a = [6.0, 2.0, 3.0]$ and $b = [1.0, 1.0, 1.0]$. Compute
the projection of $a$ onto $b$. Verify that the residual
$r = a - \text{proj}_b(a)$ is orthogonal to $b$ (i.e., $r \cdot b = 0$).

> **Hint**
>
> This orthogonality condition is the same principle behind OLS
> regression: the residuals are orthogonal to the regressors.

In [1]:
a = [6.0, 2.0, 3.0]
b = [1.0, 1.0, 1.0]

proj = (dot(a, b) / norm(b)^2) * b
println("proj_b(a) = $proj")

r = a - proj
println("Residual r = $r")
println("r ⋅ b = $(dot(r, b))")

proj_b(a) = [3.6666666666666674, 3.6666666666666674, 3.6666666666666674]
Residual r = [2.3333333333333326, -1.6666666666666674, -0.6666666666666674]
r ⋅ b = -2.220446049250313e-15

The projection onto the ones vector $[1,1,1]$ gives
$\frac{6+2+3}{3} \times [1,1,1] = [3.667, 3.667, 3.667]$, which is the
mean of the components. The residual $r = a - \text{proj}_b(a)$ is
orthogonal to $b$, confirming the geometric decomposition. This is
exactly the OLS principle: the residuals are orthogonal to the
regressors.

# Matrix Operations

## Problem 3: Input-Output Tables

Consider a simple economy with three sectors: Agriculture (A),
Manufacturing (M), and Services (S). The **input-output matrix** records
how much each sector purchases from every other sector per dollar of
output:

$$
A = \begin{pmatrix}
0.10 & 0.25 & 0.05 \\
0.20 & 0.15 & 0.10 \\
0.05 & 0.10 & 0.20
\end{pmatrix}
$$

Here $A_{ij}$ is the amount of sector $i$’s output used as input by
sector $j$ per dollar of $j$’s output.

**(a)** Create the matrix $A$ in Julia. Compute $A'$ (the transpose). Is
$A$ symmetric?

In [1]:
A = [0.10 0.25 0.05;
     0.20 0.15 0.10;
     0.05 0.10 0.20]

@show A
@show A'
println("\nIs A symmetric? $(A == A')")

A = [0.1 0.25 0.05; 0.2 0.15 0.1; 0.05 0.1 0.2]
A' = [0.1 0.2 0.05; 0.25 0.15 0.1; 0.05 0.1 0.2]

Is A symmetric? false

$A$ is not symmetric. For example, $A_{12} = 0.25$ (Agriculture’s output
used by Manufacturing) but $A_{21} = 0.20$ (Manufacturing’s output used
by Agriculture). Different sectors have different input requirements
from each other.

**(b)** Compute the column sums of $A$. What do these represent
economically? Which sector has the highest intermediate input
requirements?

> **Hint**
>
> Use `sum(A, dims=1)` to sum along rows (giving column sums).

In [1]:
col_sums = sum(A, dims=1)
println("Column sums: $col_sums")

sectors = ["Agriculture", "Manufacturing", "Services"]
for (i, s) in enumerate(sectors)
    println("$s: $(col_sums[i]) of each dollar goes to intermediate inputs")
end

Column sums: [0.35000000000000003 0.5 0.35000000000000003]
Agriculture: 0.35000000000000003 of each dollar goes to intermediate inputs
Manufacturing: 0.5 of each dollar goes to intermediate inputs
Services: 0.35000000000000003 of each dollar goes to intermediate inputs

The column sum for sector $j$ is the total intermediate input cost per
dollar of output for that sector. Manufacturing has the highest
intermediate input requirements (0.50), meaning 50 cents of every dollar
of Manufacturing output goes to purchasing inputs from other sectors.
This leaves only 50 cents per dollar for value added (wages, profits,
etc.).

**(c)** Suppose the final output vector is $x = [100, 200, 300]'$ (in
millions of dollars). Compute the total intermediate demand $Ax$.

In [1]:
x = [100.0, 200.0, 300.0]

intermediate_demand = A * x
println("Intermediate demand (Ax):")
for (i, s) in enumerate(sectors)
    println("  $s: $(intermediate_demand[i]) million")
end

Intermediate demand (Ax):
  Agriculture: 75.0 million
  Manufacturing: 80.0 million
  Services: 85.0 million

## Problem 4: Matrix Multiplication

**(a)** Create the matrices:

$$
B = \begin{pmatrix} 1 & 2 \\ 3 & 4 \end{pmatrix}, \qquad
C = \begin{pmatrix} 5 & 6 \\ 7 & 8 \end{pmatrix}
$$

Compute $BC$ and $CB$. Are they equal? Why or why not?

In [1]:
B = [1 2; 3 4]
C = [5 6; 7 8]

println("BC =")
display(B * C)
println("\nCB =")
display(C * B)
println("\nBC == CB? $(B * C == C * B)")

BC =

CB =

BC == CB? false

2×2 Matrix{Int64}:
 19  22
 43  50

2×2 Matrix{Int64}:
 23  34
 31  46

$BC \neq CB$. Matrix multiplication is **not commutative** in general.
The order matters because each product computes different linear
combinations of rows and columns. $BC$ takes linear combinations of
$C$’s rows weighted by $B$’s entries, while $CB$ takes linear
combinations of $B$’s rows weighted by $C$’s entries.

**(b)** Compute $(BC)'$ and $C'B'$. Verify that they are equal. This is
a fundamental property of matrix transposition: $(BC)' = C'B'$.

In [1]:
println("(BC)' =")
display((B * C)')
println("\nC'B' =")
display(C' * B')
println("\nEqual? $((B * C)' == C' * B')")

(BC)' =

C'B' =

Equal? true

2×2 adjoint(::Matrix{Int64}) with eltype Int64:
 19  43
 22  50

2×2 Matrix{Int64}:
 19  43
 22  50

**(c)** Compute $B \circ C$ (element-wise multiplication using `.*`).
Compare this to $BC$. Are they different?

In [1]:
println("B .* C (element-wise) =")
display(B .* C)
println("\nB * C (matrix multiplication) =")
display(B * C)
println("\nSame? $(B .* C == B * C)")

B .* C (element-wise) =

B * C (matrix multiplication) =

Same? false

2×2 Matrix{Int64}:
  5  12
 21  32

2×2 Matrix{Int64}:
 19  22
 43  50

They are different. Element-wise multiplication (`.*`) multiplies
corresponding entries: $(B \circ C)_{ij} = B_{ij} \times C_{ij}$. Matrix
multiplication computes dot products of rows and columns:
$(BC)_{ij} = \sum_k B_{ik} C_{kj}$. These are fundamentally different
operations.

# Solving Linear Systems

## Problem 5: Market Equilibrium

Consider a two-good economy. The market clearing conditions require:

$$
\begin{aligned}
2p_1 + p_2 &= 10 \\
p_1 + 3p_2 &= 15
\end{aligned}
$$

where $p_1$ and $p_2$ are the equilibrium prices.

**(a)** Write this system in matrix form $Ax = b$. Create the matrix $A$
and vector $b$ in Julia.

In [1]:
A = [2.0 1.0;
     1.0 3.0]
b = [10.0, 15.0]

println("A =")
display(A)
println("\nb = $b")

A =

b = [10.0, 15.0]

2×2 Matrix{Float64}:
 2.0  1.0
 1.0  3.0

**(b)** Verify that $A$ is invertible by computing $\det(A)$.

In [1]:
d = det(A)
println("det(A) = $d")
println("A is invertible: $(d != 0)")

det(A) = 5.0
A is invertible: true

Since $\det(A) = 2 \times 3 - 1 \times 1 = 5 \neq 0$, the matrix $A$ is
invertible and the system has a unique solution.

**(c)** Solve for the equilibrium prices using the backslash operator
`A \ b`. Verify your answer by computing $A x$ and checking that it
equals $b$.

In [1]:
p = A \ b
println("Equilibrium prices: p₁ = $(p[1]), p₂ = $(p[2])")
println("Verification: A * p = $(A * p)")
println("Matches b? $(A * p ≈ b)")

Equilibrium prices: p₁ = 3.0, p₂ = 4.0
Verification: A * p = [10.0, 15.0]
Matches b? true

**(d)** Also solve using the inverse: $x = A^{-1} b$. Verify that you
get the same answer.

In [1]:
p_inv = inv(A) * b
println("Using inverse: p₁ = $(p_inv[1]), p₂ = $(p_inv[2])")
println("Same as backslash? $(p ≈ p_inv)")

Using inverse: p₁ = 3.0, p₂ = 4.0
Same as backslash? true

## Problem 6: Leontief Model

The **Leontief input-output model** determines the total output $x$
needed to satisfy a vector of final demands $d$, accounting for the fact
that producing output requires intermediate inputs. The relationship is:

$$
x = Ax + d
$$

where $A$ is the input-output matrix from Problem 3 and $d$ is the final
demand vector. Rearranging:

$$
(I - A) x = d
$$

**(a)** Suppose final demand is $d = [50, 80, 120]'$ (millions of
dollars). Construct the matrix $(I - A)$ where $I$ is the $3 \times 3$
identity matrix. Verify that $(I - A)$ is invertible.

> **Hint**
>
> Julia provides a special identity matrix object `I` that automatically
> conforms to the correct size, so you can simply write `I - A`.

In [1]:
A = [0.10 0.25 0.05;
     0.20 0.15 0.10;
     0.05 0.10 0.20]

d = [50.0, 80.0, 120.0]

IminusA = I - A
println("I - A =")
display(IminusA)
println("\ndet(I - A) = $(det(IminusA))")
println("Invertible: $(det(IminusA) != 0)")

I - A =

det(I - A) = 0.558625
Invertible: true

3×3 Matrix{Float64}:
  0.9   -0.25  -0.05
 -0.2    0.85  -0.1
 -0.05  -0.1    0.8

**(b)** Solve for the total output vector $x = (I - A)^{-1} d$ using the
backslash operator. How much must each sector produce in total
(including intermediate goods)?

In [1]:
x = (I - A) \ d
sectors = ["Agriculture", "Manufacturing", "Services"]
println("Total output required:")
for (i, s) in enumerate(sectors)
    println("  $s: $(round(x[i], digits=2)) million")
end

Total output required:
  Agriculture: 103.83 million
  Manufacturing: 139.0 million
  Services: 173.86 million

Each sector must produce more than its final demand because it also
needs to supply intermediate inputs to the other sectors (and to
itself).

**(c)** Compute the **Leontief inverse** $L = (I - A)^{-1}$ explicitly.
The element $L_{ij}$ tells you the total output of sector $i$ required
(directly and indirectly) to deliver one dollar of final demand from
sector $j$. Which sector has the largest total output multiplier
(largest column sum of $L$)?

In [1]:
L = inv(I - A)
println("Leontief inverse L =")
display(round.(L, digits=4))

col_sums_L = sum(L, dims=1)
println("\nOutput multipliers (column sums of L):")
for (i, s) in enumerate(sectors)
    println("  $s: $(round(col_sums_L[i], digits=4))")
end

Leontief inverse L =

Output multipliers (column sums of L):
  Agriculture: 1.6066
  Manufacturing: 1.8349
  Services: 1.5798

3×3 Matrix{Float64}:
 1.1994  0.367   0.1208
 0.2954  1.2844  0.179
 0.1119  0.1835  1.2799

The sector with the largest column sum has the largest output
multiplier, meaning that a one-dollar increase in final demand for that
sector generates the largest total increase in economy-wide output
(accounting for all direct and indirect intermediate input
requirements).

# Economic Application: Input-Output Model

## Problem 10: The Leontief Model in Depth

Consider a four-sector economy (Agriculture, Manufacturing, Energy,
Services) with the following technology matrix:

$$
A = \begin{pmatrix}
0.05 & 0.15 & 0.10 & 0.02 \\
0.10 & 0.20 & 0.15 & 0.08 \\
0.08 & 0.10 & 0.05 & 0.05 \\
0.03 & 0.05 & 0.10 & 0.10
\end{pmatrix}
$$

The element $A_{ij}$ represents how many dollars of sector $i$ output
are needed to produce one dollar of sector $j$ output.

**(a)** Create the matrix $A$ in Julia. Compute the column sums. What
fraction of each sector’s gross output goes to intermediate use (as
opposed to final demand)?

In [1]:
A = [0.05 0.15 0.10 0.02;
     0.10 0.20 0.15 0.08;
     0.08 0.10 0.05 0.05;
     0.03 0.05 0.10 0.10]

sectors = ["Agriculture", "Manufacturing", "Energy", "Services"]

col_sums = sum(A, dims=1)
println("Column sums (intermediate input shares):")
for (i, s) in enumerate(sectors)
    println("  $s: $(col_sums[i]) ($(round(col_sums[i]*100, digits=1))% of gross output)")
end

Column sums (intermediate input shares):
  Agriculture: 0.26 (26.0% of gross output)
  Manufacturing: 0.49999999999999994 (50.0% of gross output)
  Energy: 0.4 (40.0% of gross output)
  Services: 0.25 (25.0% of gross output)

Manufacturing has the highest intermediate input share (50%), meaning
half of every dollar of Manufacturing output goes to purchasing
intermediate inputs. Services has the lowest (25%), reflecting that
services tend to be more labor-intensive with less reliance on material
inputs.

**(b)** Verify that the Leontief model is well-defined by checking that
all eigenvalues of $A$ are less than 1 in absolute value. This is a
necessary condition for the Leontief inverse $(I - A)^{-1}$ to exist and
have non-negative entries.

> **Hawkins-Simon Condition**
>
> If all eigenvalues of $A$ are strictly inside the unit circle
> ($|\lambda_i| < 1$), the economy is productive: it can produce
> positive net output in every sector. If any eigenvalue equals or
> exceeds 1, the economy requires more inputs than it produces.

In [1]:
eigenvalues = eigvals(A)
println("Eigenvalues of A:")
for (i, λ) in enumerate(eigenvalues)
    println("  λ$i = $(round(λ, digits=6)),  |λ$i| = $(round(abs(λ), digits=6))")
end
println("\nAll |λ| < 1? $(all(abs.(eigenvalues) .< 1))")
println("Leontief model is well-defined.")

Eigenvalues of A:
  λ1 = -0.045366,  |λ1| = 0.045366
  λ2 = -0.002521,  |λ2| = 0.002521
  λ3 = 0.069291,  |λ3| = 0.069291
  λ4 = 0.378595,  |λ4| = 0.378595

All |λ| < 1? true
Leontief model is well-defined.

All eigenvalues are strictly inside the unit circle, so the economy is
productive and the Leontief inverse exists with non-negative entries.

**(c)** Compute the Leontief inverse $L = (I - A)^{-1}$. Find the output
multiplier for each sector (the column sums of $L$). Which sector has
the largest multiplier? Interpret this economically.

In [1]:
L = inv(I - A)
println("Leontief inverse L =")
display(round.(L, digits=4))

multipliers = sum(L, dims=1)
println("\nOutput multipliers (column sums of L):")
for (i, s) in enumerate(sectors)
    println("  $s: $(round(multipliers[i], digits=4))")
end

Leontief inverse L =

Output multipliers (column sums of L):
  Agriculture: 1.4247
  Manufacturing: 1.8087
  Energy: 1.635
  Services: 1.3944

4×4 Matrix{Float64}:
 1.0914  0.2275  0.1564  0.0532
 0.1632  1.3189  0.2396  0.1342
 0.1121  0.1632  1.0984  0.078
 0.0579  0.099   0.1406  1.129

The sector with the largest multiplier generates the most total
economy-wide output per dollar of final demand. This sector has the
deepest supply chain linkages: increasing its final demand requires
substantial intermediate inputs from other sectors, which in turn
require their own inputs, creating a large cascade of production
throughout the economy.

**(d)** Suppose final demand is $d = [30, 60, 40, 100]'$ (millions of
dollars). Solve for the total output required from each sector. How much
gross output does Manufacturing need to produce?

In [1]:
d = [30.0, 60.0, 40.0, 100.0]

x = (I - A) \ d
println("Total output required:")
for (i, s) in enumerate(sectors)
    println("  $s: $(round(x[i], digits=2)) million")
end
println("\nManufacturing gross output: $(round(x[2], digits=2)) million")

Total output required:
  Agriculture: 57.97 million
  Manufacturing: 107.03 million
  Energy: 64.9 million
  Services: 126.2 million

Manufacturing gross output: 107.03 million

**(e)** **Demand shock analysis.** Now suppose final demand for Services
increases by 20% (from 100 to 120 million), while all other demands
remain unchanged. Compute the new total output vector. By how much does
each sector’s output increase? Which sector is most affected (in
absolute terms) by the increase in Services demand?

In [1]:
d_new = [30.0, 60.0, 40.0, 120.0]
x_new = (I - A) \ d_new

Δx = x_new - x
println("Output changes from 20% Services demand increase:")
for (i, s) in enumerate(sectors)
    println("  $s: +$(round(Δx[i], digits=2)) million ($(round(Δx[i]/x[i]*100, digits=1))% increase)")
end

max_idx = argmax(Δx)
println("\nMost affected sector (absolute): $(sectors[max_idx]) (+$(round(Δx[max_idx], digits=2)) million)")

Output changes from 20% Services demand increase:
  Agriculture: +1.06 million (1.8% increase)
  Manufacturing: +2.68 million (2.5% increase)
  Energy: +1.56 million (2.4% increase)
  Services: +22.58 million (17.9% increase)

Most affected sector (absolute): Services (+22.58 million)

The demand shock in Services ripples through the entire economy via
inter-industry linkages. Every sector must increase output, not just
Services, because producing more Services requires intermediate inputs
from all other sectors.

**(f)** **Structural change.** Suppose a technological improvement
reduces Manufacturing’s input requirement from Agriculture by half:
$A_{1,2}$ drops from 0.15 to 0.075. Recompute the Leontief inverse and
the output multiplier for Manufacturing. How does the technological
change affect the Manufacturing multiplier?

In [1]:
A_new = copy(A)
A_new[1, 2] = 0.075

L_new = inv(I - A_new)

multiplier_old = sum(L[:, 2])
multiplier_new = sum(L_new[:, 2])

println("Manufacturing output multiplier:")
println("  Before: $(round(multiplier_old, digits=4))")
println("  After:  $(round(multiplier_new, digits=4))")
println("  Change: $(round(multiplier_new - multiplier_old, digits=4))")

Manufacturing output multiplier:
  Before: 1.8087
  After:  1.6694
  Change: -0.1392

The technological improvement reduces the Manufacturing multiplier. When
Manufacturing requires less Agricultural input per dollar of output, the
total cascade of inter-industry production triggered by a dollar of
Manufacturing final demand is smaller. This is a productivity gain: the
economy can satisfy the same final demand with less total gross output.